api/src/play_asset_pack_manager.gd

```diff
@@ -105,9 +124,103 @@ func _initialize_plugin() -> Object:
+# _asset_pack_to_request_map.
+# -----------------------------------------------------------------------------
+func _remove_request_reference_from_map(pack_name : String):
+	_play_asset_pack_manager_mutex.lock()	
+	_asset_pack_to_request_map.erase(pack_name)
+	_play_asset_pack_manager_mutex.unlock()	
+
+# -----------------------------------------------------------------------------
+# Helper function that synchronizes relevant request object's state upon 
+# receiving assetPackStateUpdated signal.
+# -----------------------------------------------------------------------------
+func _route_asset_pack_state_updated(result : Dictionary):
+	var updated_state : PlayAssetPackState = PlayAssetPackState.new(result)
+	var pack_name = updated_state.get_name()
+	var updated_status = updated_state.get_status()
+	
+	_play_asset_pack_manager_mutex.lock()	
+	
+	var pack_name_exists_in_request_map = _asset_pack_to_request_map.has(pack_name)
+	var received_fetch_callback
+	var duplicate_state
+	if pack_name_exists_in_request_map:
+		var request = _asset_pack_to_request_map[pack_name]
+		received_fetch_callback = request.get_state().get_status() != PlayAssetPackManager.AssetPackStatus.UNKNOWN and\
+			request.get_state().get_total_bytes_to_download() != 0
+		duplicate_state = request.get_state().to_dict().hash() == result.hash()
+		
+		# Since assetPackStateUpdated and fetchSuccess/Error might contain duplicate updates,
+		# we will only route non-duplicate updates to request object after we already received 
+		# fetchSuccess/Error signal.
+		if received_fetch_callback and not duplicate_state:
+			# Since devs might read request's state while we are updating it, we need to call this	
+			# function from main thread.
+			request.call_deferred("_on_state_updated", result)		
+		
+		# if reached terminal state, release references	
+		if updated_status in _PACK_TERMINAL_STATES:	
+			_asset_pack_to_request_map.erase(pack_name)
+		
+	# only emit non-repeated state_updated signals after we encountered fetchSuccess/Error
+	if pack_name_exists_in_request_map:
+		if received_fetch_callback and not duplicate_state:
+			_asset_pack_state_cache[pack_name] = result
+			# emit state updated signal on main thread
+			call_deferred("emit_signal", "state_updated", pack_name, updated_state)
+	else:
+		_asset_pack_state_cache[pack_name] = result
+		call_deferred("emit_signal", "state_updated", pack_name, updated_state)	
+	
+	_play_asset_pack_manager_mutex.unlock()
+
+# -----------------------------------------------------------------------------
+# Helper function called by request objects, to emit artifical state_updated signals.
+# -----------------------------------------------------------------------------
+func _forward_high_level_state_updated_signal(pack_name : String, state : Dictionary):
+	# update cache, since this function can be called by multiple request objects with same packName
+	_play_asset_pack_manager_mutex.lock()
```

Comment:
I don't know if you need the mutex here since changing dictionary values is thread safe.

Reply:
Removed mutex in this function.